pandas/core/case_when.py

```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations
+
+from typing import (
+    Any,
+    Callable,
+)
+
+import numpy as np
+
+from pandas._libs import lib
+
+import pandas as pd
+import pandas.core.common as com
+
+
+def case_when(*args, default: Any = lib.no_default) -> Callable:
+    """
+    Create a callable for assignment based on a condition or multiple conditions.
+
+    This is useful when you want to assign a column based on multiple conditions.
+
+    Parameters
+    ----------
+    args : Variable argument of conditions and expected values.
+        Takes the form:
+            `condition0`, `value0`, `condition1`, `value1`, ...
+        `condition` can be a 1-D boolean array/series or a callable
+        that evaluate to a 1-D boolean array/series.
+    default : Any, default is `None`.
+        The default value to be used if all conditions evaluate False.
```

Comment:
If no value is specified, would we not want the default for it to remain untouched?

When default is not specified, I believe the current implementation will convert e.g. int to object dtypes. That seems undesirable to me.

Reply:
Indeed. For now, we can use `np.nan`. Pushed a fixup: [5595161](https://github.com/pandas-dev/pandas/pull/50343/commits/559516128b09567b623626119f63afad19f9e05b)

Using `np.nan` also preserves date related dtypes properly. To illustrate: 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3]))
>>> result = df.assign(
...     new_column=pd.case_when(
...         lambda x: x.a == 1, pd.Timestamp("today")
...     )
... )
>>> result.new_column
0   2023-01-07 18:58:28.654455
1                          NaT
2                          NaT
Name: new_column, dtype: datetime64[ns]
```

Reply:
I was in the mindset that we were doing something like

```
ser.mask(cond1, value1).mask(~cond1 & cond2, value2)
```

with a starting Series. You can indeed get this behavior with case_when by specifying `.case_when(..., default=ser)`. So I think behavior-wise this is good.

However I don't think the default of np.nan will work for all dtypes, e.g. in my experience it's common (but not universal) to have None instead of np.nan be the NA value for string. Leaving it as np.nan might unexpectedly coerce the dtype on some inputs as well, surprising users.

With this, it seems better to me to have `default` be a required argument.

Reply:
> With this, it seems better to me to have default be a required argument.

Good idea, done. 

> Also, how does the performance compare between:

Is there a need to check this performance as now `case_when` uses `Series.mask`?